File: templates/problems/01_arrays_hashing/02_valid_anagram/valid_anagram_test.go
package valid_anagram

import "testing"

func TestIsAnagram(t *testing.T) {
	tests := []struct {
		name string
		s    string
		t    string
		want bool
	}{
		{
			name: "valid anagram",
			s:    "anagram",
			t:    "nagaram",
			want: true,
		},
		{
			name: "not anagram",
			s:    "rat",
			t:    "car",
			want: false,
		},
		{
			name: "empty strings",
			s:    "",
			t:    "",
			want: true,
		},
		{
			name: "different lengths",
			s:    "ab",
			t:    "abc",
			want: false,
		},
		{
			name: "single characters same",
			s:    "a",
			t:    "a",
			want: true,
		},
		{
			name: "single characters different",
			s:    "a",
			t:    "b",
			want: false,
		},
		{
			name: "repeated characters",
			s:    "aabb",
			t:    "bbaa",
			want: true,
		},
		{
			name: "same chars different frequency",
			s:    "aaab",
			t:    "aabb",
			want: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := isAnagram(tt.s, tt.t)
			if got != tt.want {
				t.Errorf("isAnagram(%q, %q) = %v, want %v", tt.s, tt.t, got, tt.want)
			}
		})
	}
}

File: templates/problems/01_arrays_hashing/02_valid_anagram/typescript/validAnagram.test.ts
import { describe, it, expect } from 'vitest';
import { isAnagram } from './validAnagram';

describe('ValidAnagram', () => {
    it('valid anagram', () => {
        expect(isAnagram('anagram', 'nagaram')).toBe(true);
    });

    it('not anagram', () => {
        expect(isAnagram('rat', 'car')).toBe(false);
    });

    it('empty strings', () => {
        expect(isAnagram('', '')).toBe(true);
    });

    it('different lengths', () => {
        expect(isAnagram('ab', 'abc')).toBe(false);
    });

    it('single characters same', () => {
        expect(isAnagram('a', 'a')).toBe(true);
    });

    it('single characters different', () => {
        expect(isAnagram('a', 'b')).toBe(false);
    });

    it('repeated characters', () => {
        expect(isAnagram('aabb', 'bbaa')).toBe(true);
    });

    it('same chars different frequency', () => {
        expect(isAnagram('aaab', 'aabb')).toBe(false);
    });
});
